// @acala-network/chopsticks-executor/dist/cjs/chopsticks_executor.js
/* @ts-self-types="./chopsticks_executor.d.ts" */ /**
 * @param {any} entries
 * @param {any} trie_version
 * @returns {Promise<any>}
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get calculate_state_root () {
        return calculate_state_root;
    },
    get create_proof () {
        return create_proof;
    },
    get decode_proof () {
        return decode_proof;
    },
    get default () {
        return __wbg_init;
    },
    get get_runtime_version () {
        return get_runtime_version;
    },
    get initSync () {
        return initSync;
    },
    get run_task () {
        return run_task;
    },
    get testing () {
        return testing;
    }
});
function _array_like_to_array(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _array_without_holes(arr) {
    if (Array.isArray(arr)) return _array_like_to_array(arr);
}
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self1 = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self1, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _instanceof(left, right) {
    if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
        return !!right[Symbol.hasInstance](left);
    } else {
        return left instanceof right;
    }
}
function _iterable_to_array(iter) {
    if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter);
}
function _non_iterable_spread() {
    throw new TypeError("Invalid attempt to spread non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _to_consumable_array(arr) {
    return _array_without_holes(arr) || _iterable_to_array(arr) || _unsupported_iterable_to_array(arr) || _non_iterable_spread();
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _unsupported_iterable_to_array(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
}
function _ts_generator(thisArg, body) {
    var f, y, t, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    }, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype), d = Object.defineProperty;
    return d(g, "next", {
        value: verb(0)
    }), d(g, "throw", {
        value: verb(1)
    }), d(g, "return", {
        value: verb(2)
    }), typeof Symbol === "function" && d(g, Symbol.iterator, {
        value: function() {
            return this;
        }
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(g && (g = 0, op[0] && (_ = 0)), _)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
function calculate_state_root(entries, trie_version) {
    var ret = wasm.calculate_state_root(entries, trie_version);
    return ret;
}
function create_proof(nodes, updates) {
    var ret = wasm.create_proof(nodes, updates);
    return ret;
}
function decode_proof(trie_root_hash, nodes) {
    var ret = wasm.decode_proof(trie_root_hash, nodes);
    return ret;
}
function get_runtime_version(code) {
    var ret = wasm.get_runtime_version(code);
    return ret;
}
function run_task(task, js) {
    var ret = wasm.run_task(task, js);
    return ret;
}
function testing(js, key) {
    var ret = wasm.testing(js, key);
    return ret;
}
function __wbg_get_imports() {
    var import0 = {
        __proto__: null,
        __wbg_Error_8c4e43fe74559d73: function __wbg_Error_8c4e43fe74559d73(arg0, arg1) {
            var ret = Error(getStringFromWasm0(arg0, arg1));
            return ret;
        },
        __wbg_Number_04624de7d0e8332d: function __wbg_Number_04624de7d0e8332d(arg0) {
            var ret = Number(arg0);
            return ret;
        },
        __wbg_String_8f0eb39a4a4c2f66: function __wbg_String_8f0eb39a4a4c2f66(arg0, arg1) {
            var ret = String(arg1);
            var ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg___wbindgen_bigint_get_as_i64_8fcf4ce7f1ca72a2: function __wbg___wbindgen_bigint_get_as_i64_8fcf4ce7f1ca72a2(arg0, arg1) {
            var v = arg1;
            var ret = (typeof v === "undefined" ? "undefined" : _type_of(v)) === 'bigint' ? v : undefined;
            getDataViewMemory0().setBigInt64(arg0 + 8 * 1, isLikeNone(ret) ? BigInt(0) : ret, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, !isLikeNone(ret), true);
        },
        __wbg___wbindgen_boolean_get_bbbb1c18aa2f5e25: function __wbg___wbindgen_boolean_get_bbbb1c18aa2f5e25(arg0) {
            var v = arg0;
            var ret = typeof v === 'boolean' ? v : undefined;
            return isLikeNone(ret) ? 0xFFFFFF : ret ? 1 : 0;
        },
        __wbg___wbindgen_debug_string_0bc8482c6e3508ae: function __wbg___wbindgen_debug_string_0bc8482c6e3508ae(arg0, arg1) {
            var ret = debugString(arg1);
            var ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg___wbindgen_in_47fa6863be6f2f25: function __wbg___wbindgen_in_47fa6863be6f2f25(arg0, arg1) {
            var ret = arg0 in arg1;
            return ret;
        },
        __wbg___wbindgen_is_bigint_31b12575b56f32fc: function __wbg___wbindgen_is_bigint_31b12575b56f32fc(arg0) {
            var ret = (typeof arg0 === "undefined" ? "undefined" : _type_of(arg0)) === 'bigint';
            return ret;
        },
        __wbg___wbindgen_is_function_0095a73b8b156f76: function __wbg___wbindgen_is_function_0095a73b8b156f76(arg0) {
            var ret = typeof arg0 === 'function';
            return ret;
        },
        __wbg___wbindgen_is_object_5ae8e5880f2c1fbd: function __wbg___wbindgen_is_object_5ae8e5880f2c1fbd(arg0) {
            var val = arg0;
            var ret = (typeof val === "undefined" ? "undefined" : _type_of(val)) === 'object' && val !== null;
            return ret;
        },
        __wbg___wbindgen_is_string_cd444516edc5b180: function __wbg___wbindgen_is_string_cd444516edc5b180(arg0) {
            var ret = typeof arg0 === 'string';
            return ret;
        },
        __wbg___wbindgen_is_undefined_9e4d92534c42d778: function __wbg___wbindgen_is_undefined_9e4d92534c42d778(arg0) {
            var ret = arg0 === undefined;
            return ret;
        },
        __wbg___wbindgen_jsval_eq_11888390b0186270: function __wbg___wbindgen_jsval_eq_11888390b0186270(arg0, arg1) {
            var ret = arg0 === arg1;
            return ret;
        },
        __wbg___wbindgen_jsval_loose_eq_9dd77d8cd6671811: function __wbg___wbindgen_jsval_loose_eq_9dd77d8cd6671811(arg0, arg1) {
            var ret = arg0 == arg1;
            return ret;
        },
        __wbg___wbindgen_number_get_8ff4255516ccad3e: function __wbg___wbindgen_number_get_8ff4255516ccad3e(arg0, arg1) {
            var obj = arg1;
            var ret = typeof obj === 'number' ? obj : undefined;
            getDataViewMemory0().setFloat64(arg0 + 8 * 1, isLikeNone(ret) ? 0 : ret, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, !isLikeNone(ret), true);
        },
        __wbg___wbindgen_string_get_72fb696202c56729: function __wbg___wbindgen_string_get_72fb696202c56729(arg0, arg1) {
            var obj = arg1;
            var ret = typeof obj === 'string' ? obj : undefined;
            var ptr1 = isLikeNone(ret) ? 0 : passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg___wbindgen_throw_be289d5034ed271b: function __wbg___wbindgen_throw_be289d5034ed271b(arg0, arg1) {
            throw new Error(getStringFromWasm0(arg0, arg1));
        },
        __wbg__wbg_cb_unref_d9b87ff7982e3b21: function __wbg__wbg_cb_unref_d9b87ff7982e3b21(arg0) {
            arg0._wbg_cb_unref();
        },
        __wbg_call_389efe28435a9388: function __wbg_call_389efe28435a9388() {
            return handleError(function(arg0, arg1) {
                var ret = arg0.call(arg1);
                return ret;
            }, arguments);
        },
        __wbg_call_4708e0c13bdc8e95: function __wbg_call_4708e0c13bdc8e95() {
            return handleError(function(arg0, arg1, arg2) {
                var ret = arg0.call(arg1, arg2);
                return ret;
            }, arguments);
        },
        __wbg_debug_a4099fa12db6cd61: function __wbg_debug_a4099fa12db6cd61(arg0) {
            console.debug(arg0);
        },
        __wbg_done_57b39ecd9addfe81: function __wbg_done_57b39ecd9addfe81(arg0) {
            var ret = arg0.done;
            return ret;
        },
        __wbg_error_7534b8e9a36f1ab4: function __wbg_error_7534b8e9a36f1ab4(arg0, arg1) {
            var deferred0_0;
            var deferred0_1;
            try {
                deferred0_0 = arg0;
                deferred0_1 = arg1;
                console.error(getStringFromWasm0(arg0, arg1));
            } finally{
                wasm.__wbindgen_free(deferred0_0, deferred0_1, 1);
            }
        },
        __wbg_error_9a7fe3f932034cde: function __wbg_error_9a7fe3f932034cde(arg0) {
            console.error(arg0);
        },
        __wbg_getNextKey_1bc9de7af3b8ddff: function __wbg_getNextKey_1bc9de7af3b8ddff() {
            return handleError(function(arg0, arg1, arg2) {
                var ret = arg0.getNextKey(arg1, arg2);
                return ret;
            }, arguments);
        },
        __wbg_getStorage_07fca860094dc7ef: function __wbg_getStorage_07fca860094dc7ef() {
            return handleError(function(arg0, arg1) {
                var ret = arg0.getStorage(arg1);
                return ret;
            }, arguments);
        },
        __wbg_get_9b94d73e6221f75c: function __wbg_get_9b94d73e6221f75c(arg0, arg1) {
            var ret = arg0[arg1 >>> 0];
            return ret;
        },
        __wbg_get_b3ed3ad4be2bc8ac: function __wbg_get_b3ed3ad4be2bc8ac() {
            return handleError(function(arg0, arg1) {
                var ret = Reflect.get(arg0, arg1);
                return ret;
            }, arguments);
        },
        __wbg_get_with_ref_key_1dc361bd10053bfe: function __wbg_get_with_ref_key_1dc361bd10053bfe(arg0, arg1) {
            var ret = arg0[arg1];
            return ret;
        },
        __wbg_info_148d043840582012: function __wbg_info_148d043840582012(arg0) {
            console.info(arg0);
        },
        __wbg_instanceof_ArrayBuffer_c367199e2fa2aa04: function __wbg_instanceof_ArrayBuffer_c367199e2fa2aa04(arg0) {
            var result;
            try {
                result = _instanceof(arg0, ArrayBuffer);
            } catch (_) {
                result = false;
            }
            var ret = result;
            return ret;
        },
        __wbg_instanceof_Uint8Array_9b9075935c74707c: function __wbg_instanceof_Uint8Array_9b9075935c74707c(arg0) {
            var result;
            try {
                result = _instanceof(arg0, Uint8Array);
            } catch (_) {
                result = false;
            }
            var ret = result;
            return ret;
        },
        __wbg_isArray_d314bb98fcf08331: function __wbg_isArray_d314bb98fcf08331(arg0) {
            var ret = Array.isArray(arg0);
            return ret;
        },
        __wbg_isSafeInteger_bfbc7332a9768d2a: function __wbg_isSafeInteger_bfbc7332a9768d2a(arg0) {
            var ret = Number.isSafeInteger(arg0);
            return ret;
        },
        __wbg_iterator_6ff6560ca1568e55: function __wbg_iterator_6ff6560ca1568e55() {
            var ret = Symbol.iterator;
            return ret;
        },
        __wbg_length_32ed9a279acd054c: function __wbg_length_32ed9a279acd054c(arg0) {
            var ret = arg0.length;
            return ret;
        },
        __wbg_length_35a7bace40f36eac: function __wbg_length_35a7bace40f36eac(arg0) {
            var ret = arg0.length;
            return ret;
        },
        __wbg_log_6b5ca2e6124b2808: function __wbg_log_6b5ca2e6124b2808(arg0) {
            console.log(arg0);
        },
        __wbg_new_361308b2356cecd0: function __wbg_new_361308b2356cecd0() {
            var ret = new Object();
            return ret;
        },
        __wbg_new_3eb36ae241fe6f44: function __wbg_new_3eb36ae241fe6f44() {
            var ret = new Array();
            return ret;
        },
        __wbg_new_8a6f238a6ece86ea: function __wbg_new_8a6f238a6ece86ea() {
            var ret = new Error();
            return ret;
        },
        __wbg_new_b5d9e2fb389fef91: function __wbg_new_b5d9e2fb389fef91(arg0, arg1) {
            try {
                var state0 = {
                    a: arg0,
                    b: arg1
                };
                var cb0 = function cb0(arg0, arg1) {
                    var a = state0.a;
                    state0.a = 0;
                    try {
                        return wasm_bindgen__convert__closures_____invoke__h0dafe9b2bb334f52(a, state0.b, arg0, arg1);
                    } finally{
                        state0.a = a;
                    }
                };
                var ret = new Promise(cb0);
                return ret;
            } finally{
                state0.a = state0.b = 0;
            }
        },
        __wbg_new_dd2b680c8bf6ae29: function __wbg_new_dd2b680c8bf6ae29(arg0) {
            var ret = new Uint8Array(arg0);
            return ret;
        },
        __wbg_new_no_args_1c7c842f08d00ebb: function __wbg_new_no_args_1c7c842f08d00ebb(arg0, arg1) {
            var ret = new Function(getStringFromWasm0(arg0, arg1));
            return ret;
        },
        __wbg_next_3482f54c49e8af19: function __wbg_next_3482f54c49e8af19() {
            return handleError(function(arg0) {
                var ret = arg0.next();
                return ret;
            }, arguments);
        },
        __wbg_next_418f80d8f5303233: function __wbg_next_418f80d8f5303233(arg0) {
            var ret = arg0.next;
            return ret;
        },
        __wbg_offchainGetStorage_cfd64afb534b2e44: function __wbg_offchainGetStorage_cfd64afb534b2e44() {
            return handleError(function(arg0, arg1) {
                var ret = arg0.offchainGetStorage(arg1);
                return ret;
            }, arguments);
        },
        __wbg_offchainRandomSeed_a96103eb1425293e: function __wbg_offchainRandomSeed_a96103eb1425293e() {
            return handleError(function(arg0) {
                var ret = arg0.offchainRandomSeed();
                return ret;
            }, arguments);
        },
        __wbg_offchainSubmitTransaction_a96e65173ea41b5e: function __wbg_offchainSubmitTransaction_a96e65173ea41b5e() {
            return handleError(function(arg0, arg1) {
                var ret = arg0.offchainSubmitTransaction(arg1);
                return ret;
            }, arguments);
        },
        __wbg_offchainTimestamp_0e234957c6dcda00: function __wbg_offchainTimestamp_0e234957c6dcda00() {
            return handleError(function(arg0) {
                var ret = arg0.offchainTimestamp();
                return ret;
            }, arguments);
        },
        __wbg_prototypesetcall_bdcdcc5842e4d77d: function __wbg_prototypesetcall_bdcdcc5842e4d77d(arg0, arg1, arg2) {
            Uint8Array.prototype.set.call(getArrayU8FromWasm0(arg0, arg1), arg2);
        },
        __wbg_queueMicrotask_0aa0a927f78f5d98: function __wbg_queueMicrotask_0aa0a927f78f5d98(arg0) {
            var ret = arg0.queueMicrotask;
            return ret;
        },
        __wbg_queueMicrotask_5bb536982f78a56f: function __wbg_queueMicrotask_5bb536982f78a56f(arg0) {
            queueMicrotask(arg0);
        },
        __wbg_resolve_002c4b7d9d8f6b64: function __wbg_resolve_002c4b7d9d8f6b64(arg0) {
            var ret = Promise.resolve(arg0);
            return ret;
        },
        __wbg_set_3f1d0b984ed272ed: function __wbg_set_3f1d0b984ed272ed(arg0, arg1, arg2) {
            arg0[arg1] = arg2;
        },
        __wbg_set_f43e577aea94465b: function __wbg_set_f43e577aea94465b(arg0, arg1, arg2) {
            arg0[arg1 >>> 0] = arg2;
        },
        __wbg_stack_0ed75d68575b0f3c: function __wbg_stack_0ed75d68575b0f3c(arg0, arg1) {
            var ret = arg1.stack;
            var ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_static_accessor_GLOBAL_12837167ad935116: function __wbg_static_accessor_GLOBAL_12837167ad935116() {
            var ret = typeof global === 'undefined' ? null : global;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_static_accessor_GLOBAL_THIS_e628e89ab3b1c95f: function __wbg_static_accessor_GLOBAL_THIS_e628e89ab3b1c95f() {
            var ret = typeof globalThis === 'undefined' ? null : globalThis;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_static_accessor_SELF_a621d3dfbb60d0ce: function __wbg_static_accessor_SELF_a621d3dfbb60d0ce() {
            var ret = typeof self === 'undefined' ? null : self;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_static_accessor_WINDOW_f8727f0cf888e0bd: function __wbg_static_accessor_WINDOW_f8727f0cf888e0bd() {
            var ret = typeof window === 'undefined' ? null : window;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_then_0d9fe2c7b1857d32: function __wbg_then_0d9fe2c7b1857d32(arg0, arg1, arg2) {
            var ret = arg0.then(arg1, arg2);
            return ret;
        },
        __wbg_then_b9e7b3b5f1a9e1b5: function __wbg_then_b9e7b3b5f1a9e1b5(arg0, arg1) {
            var ret = arg0.then(arg1);
            return ret;
        },
        __wbg_value_0546255b415e96c1: function __wbg_value_0546255b415e96c1(arg0) {
            var ret = arg0.value;
            return ret;
        },
        __wbg_warn_f7ae1b2e66ccb930: function __wbg_warn_f7ae1b2e66ccb930(arg0) {
            console.warn(arg0);
        },
        __wbindgen_cast_0000000000000001: function __wbindgen_cast_0000000000000001(arg0, arg1) {
            // Cast intrinsic for `Closure(Closure { dtor_idx: 139, function: Function { arguments: [Externref], shim_idx: 140, ret: Unit, inner_ret: Some(Unit) }, mutable: true }) -> Externref`.
            var ret = makeMutClosure(arg0, arg1, wasm.wasm_bindgen__closure__destroy__hb454d70b27c07de9, wasm_bindgen__convert__closures_____invoke__h374ab0d2ab71a4a8);
            return ret;
        },
        __wbindgen_cast_0000000000000002: function __wbindgen_cast_0000000000000002(arg0) {
            // Cast intrinsic for `F64 -> Externref`.
            var ret = arg0;
            return ret;
        },
        __wbindgen_cast_0000000000000003: function __wbindgen_cast_0000000000000003(arg0, arg1) {
            // Cast intrinsic for `Ref(String) -> Externref`.
            var ret = getStringFromWasm0(arg0, arg1);
            return ret;
        },
        __wbindgen_cast_0000000000000004: function __wbindgen_cast_0000000000000004(arg0) {
            // Cast intrinsic for `U64 -> Externref`.
            var ret = BigInt.asUintN(64, arg0);
            return ret;
        },
        __wbindgen_init_externref_table: function __wbindgen_init_externref_table() {
            var table = wasm.__wbindgen_externrefs;
            var offset = table.grow(4);
            table.set(0, undefined);
            table.set(offset + 0, undefined);
            table.set(offset + 1, null);
            table.set(offset + 2, true);
            table.set(offset + 3, false);
        }
    };
    return {
        __proto__: null,
        "./chopsticks_executor_bg.js": import0
    };
}
function wasm_bindgen__convert__closures_____invoke__h374ab0d2ab71a4a8(arg0, arg1, arg2) {
    wasm.wasm_bindgen__convert__closures_____invoke__h374ab0d2ab71a4a8(arg0, arg1, arg2);
}
function wasm_bindgen__convert__closures_____invoke__h0dafe9b2bb334f52(arg0, arg1, arg2, arg3) {
    wasm.wasm_bindgen__convert__closures_____invoke__h0dafe9b2bb334f52(arg0, arg1, arg2, arg3);
}
function addToExternrefTable0(obj) {
    var idx = wasm.__externref_table_alloc();
    wasm.__wbindgen_externrefs.set(idx, obj);
    return idx;
}
var CLOSURE_DTORS = typeof FinalizationRegistry === 'undefined' ? {
    register: function register() {},
    unregister: function unregister() {}
} : new FinalizationRegistry(function(state) {
    return state.dtor(state.a, state.b);
});
function debugString(val) {
    // primitive types
    var type = typeof val === "undefined" ? "undefined" : _type_of(val);
    if (type == 'number' || type == 'boolean' || val == null) {
        return "".concat(val);
    }
    if (type == 'string') {
        return '"'.concat(val, '"');
    }
    if (type == 'symbol') {
        var description = val.description;
        if (description == null) {
            return 'Symbol';
        } else {
            return "Symbol(".concat(description, ")");
        }
    }
    if (type == 'function') {
        var name = val.name;
        if (typeof name == 'string' && name.length > 0) {
            return "Function(".concat(name, ")");
        } else {
            return 'Function';
        }
    }
    // objects
    if (Array.isArray(val)) {
        var length = val.length;
        var debug = '[';
        if (length > 0) {
            debug += debugString(val[0]);
        }
        for(var i = 1; i < length; i++){
            debug += ', ' + debugString(val[i]);
        }
        debug += ']';
        return debug;
    }
    // Test for built-in
    var builtInMatches = /\[object ([^\]]+)\]/.exec(toString.call(val));
    var className;
    if (builtInMatches && builtInMatches.length > 1) {
        className = builtInMatches[1];
    } else {
        // Failed to match the standard '[object ClassName]'
        return toString.call(val);
    }
    if (className == 'Object') {
        // we're a user defined class or Object
        // JSON.stringify avoids problems with cycles, and is generally much
        // easier than looping through ownProperties of `val`.
        try {
            return 'Object(' + JSON.stringify(val) + ')';
        } catch (_) {
            return 'Object';
        }
    }
    // errors
    if (_instanceof(val, Error)) {
        return "".concat(val.name, ": ").concat(val.message, "\n").concat(val.stack);
    }
    // TODO we could test for more things here, like `Set`s and `Map`s.
    return className;
}
function getArrayU8FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getUint8ArrayMemory0().subarray(ptr / 1, ptr / 1 + len);
}
var cachedDataViewMemory0 = null;
function getDataViewMemory0() {
    if (cachedDataViewMemory0 === null || cachedDataViewMemory0.buffer.detached === true || cachedDataViewMemory0.buffer.detached === undefined && cachedDataViewMemory0.buffer !== wasm.memory.buffer) {
        cachedDataViewMemory0 = new DataView(wasm.memory.buffer);
    }
    return cachedDataViewMemory0;
}
function getStringFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return decodeText(ptr, len);
}
var cachedUint8ArrayMemory0 = null;
function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}
function handleError(f, args) {
    try {
        return f.apply(this, args);
    } catch (e) {
        var idx = addToExternrefTable0(e);
        wasm.__wbindgen_exn_store(idx);
    }
}
function isLikeNone(x) {
    return x === undefined || x === null;
}
function makeMutClosure(arg0, arg1, dtor, f) {
    var state = {
        a: arg0,
        b: arg1,
        cnt: 1,
        dtor: dtor
    };
    var real = function real1() {
        for(var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++){
            args[_key] = arguments[_key];
        }
        // First up with a closure we increment the internal reference
        // count. This ensures that the Rust closure environment won't
        // be deallocated while we're invoking it.
        state.cnt++;
        var a = state.a;
        state.a = 0;
        try {
            return f.apply(void 0, [
                a,
                state.b
            ].concat(_to_consumable_array(args)));
        } finally{
            state.a = a;
            real._wbg_cb_unref();
        }
    };
    real._wbg_cb_unref = function() {
        if (--state.cnt === 0) {
            state.dtor(state.a, state.b);
            state.a = 0;
            CLOSURE_DTORS.unregister(state);
        }
    };
    CLOSURE_DTORS.register(real, state, state);
    return real;
}
function passStringToWasm0(arg, malloc, realloc) {
    if (realloc === undefined) {
        var buf = cachedTextEncoder.encode(arg);
        var ptr = malloc(buf.length, 1) >>> 0;
        getUint8ArrayMemory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }
    var len = arg.length;
    var ptr1 = malloc(len, 1) >>> 0;
    var mem = getUint8ArrayMemory0();
    var offset = 0;
    for(; offset < len; offset++){
        var code = arg.charCodeAt(offset);
        if (code > 0x7F) break;
        mem[ptr1 + offset] = code;
    }
    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr1 = realloc(ptr1, len, len = offset + arg.length * 3, 1) >>> 0;
        var view = getUint8ArrayMemory0().subarray(ptr1 + offset, ptr1 + len);
        var ret = cachedTextEncoder.encodeInto(arg, view);
        offset += ret.written;
        ptr1 = realloc(ptr1, len, offset, 1) >>> 0;
    }
    WASM_VECTOR_LEN = offset;
    return ptr1;
}
var cachedTextDecoder = new TextDecoder('utf-8', {
    ignoreBOM: true,
    fatal: true
});
cachedTextDecoder.decode();
var MAX_SAFARI_DECODE_BYTES = 2146435072;
var numBytesDecoded = 0;
function decodeText(ptr, len) {
    numBytesDecoded += len;
    if (numBytesDecoded >= MAX_SAFARI_DECODE_BYTES) {
        cachedTextDecoder = new TextDecoder('utf-8', {
            ignoreBOM: true,
            fatal: true
        });
        cachedTextDecoder.decode();
        numBytesDecoded = len;
    }
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}
var cachedTextEncoder = new TextEncoder();
if (!('encodeInto' in cachedTextEncoder)) {
    cachedTextEncoder.encodeInto = function(arg, view) {
        var buf = cachedTextEncoder.encode(arg);
        view.set(buf);
        return {
            read: arg.length,
            written: buf.length
        };
    };
}
var WASM_VECTOR_LEN = 0;
var wasmModule, wasm;
function __wbg_finalize_init(instance, module) {
    wasm = instance.exports;
    wasmModule = module;
    cachedDataViewMemory0 = null;
    cachedUint8ArrayMemory0 = null;
    wasm.__wbindgen_start();
    return wasm;
}
function __wbg_load(module, imports) {
    return _async_to_generator(function() {
        var e, validResponse, bytes, instance;
        function expectedResponseType(type) {
            switch(type){
                case 'basic':
                case 'cors':
                case 'default':
                    return true;
            }
            return false;
        }
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    if (!(typeof Response === 'function' && _instanceof(module, Response))) return [
                        3,
                        7
                    ];
                    if (!(typeof WebAssembly.instantiateStreaming === 'function')) return [
                        3,
                        4
                    ];
                    _state.label = 1;
                case 1:
                    _state.trys.push([
                        1,
                        3,
                        ,
                        4
                    ]);
                    return [
                        4,
                        WebAssembly.instantiateStreaming(module, imports)
                    ];
                case 2:
                    return [
                        2,
                        _state.sent()
                    ];
                case 3:
                    e = _state.sent();
                    validResponse = module.ok && expectedResponseType(module.type);
                    if (validResponse && module.headers.get('Content-Type') !== 'application/wasm') {
                        console.warn("`WebAssembly.instantiateStreaming` failed because your server does not serve Wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:\n", e);
                    } else {
                        throw e;
                    }
                    return [
                        3,
                        4
                    ];
                case 4:
                    return [
                        4,
                        module.arrayBuffer()
                    ];
                case 5:
                    bytes = _state.sent();
                    return [
                        4,
                        WebAssembly.instantiate(bytes, imports)
                    ];
                case 6:
                    return [
                        2,
                        _state.sent()
                    ];
                case 7:
                    return [
                        4,
                        WebAssembly.instantiate(module, imports)
                    ];
                case 8:
                    instance = _state.sent();
                    if (_instanceof(instance, WebAssembly.Instance)) {
                        return [
                            2,
                            {
                                instance: instance,
                                module: module
                            }
                        ];
                    } else {
                        return [
                            2,
                            instance
                        ];
                    }
                    _state.label = 9;
                case 9:
                    return [
                        2
                    ];
            }
        });
    })();
}
function initSync(module) {
    if (wasm !== undefined) return wasm;
    if (module !== undefined) {
        if (Object.getPrototypeOf(module) === Object.prototype) {
            module = module.module;
        } else {
            console.warn('using deprecated parameters for `initSync()`; pass a single object instead');
        }
    }
    var imports = __wbg_get_imports();
    if (!_instanceof(module, WebAssembly.Module)) {
        module = new WebAssembly.Module(module);
    }
    var instance = new WebAssembly.Instance(module, imports);
    return __wbg_finalize_init(instance, module);
}
function __wbg_init(module_or_path) {
    return _async_to_generator(function() {
        var imports, _ref, instance, module;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    if (wasm !== undefined) return [
                        2,
                        wasm
                    ];
                    if (module_or_path !== undefined) {
                        if (Object.getPrototypeOf(module_or_path) === Object.prototype) {
                            module_or_path = module_or_path.module_or_path;
                        } else {
                            console.warn('using deprecated parameters for the initialization function; pass a single object instead');
                        }
                    }
                    if (module_or_path === undefined) {
                        module_or_path = new URL('chopsticks_executor_bg.wasm', require("url").pathToFileURL(__filename).toString());
                    }
                    imports = __wbg_get_imports();
                    if (typeof module_or_path === 'string' || typeof Request === 'function' && _instanceof(module_or_path, Request) || typeof URL === 'function' && _instanceof(module_or_path, URL)) {
                        module_or_path = fetch(module_or_path);
                    }
                    return [
                        4,
                        module_or_path
                    ];
                case 1:
                    return [
                        4,
                        __wbg_load.apply(void 0, [
                            _state.sent(),
                            imports
                        ])
                    ];
                case 2:
                    _ref = _state.sent(), instance = _ref.instance, module = _ref.module;
                    return [
                        2,
                        __wbg_finalize_init(instance, module)
                    ];
            }
        });
    })();
}
