// @acala-network/chopsticks-executor/dist/cjs/chopsticks_executor_bg.wasm.d.js
/* tslint:disable */ /* eslint-disable */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get __externref_table_alloc () {
        return __externref_table_alloc;
    },
    get __wbindgen_exn_store () {
        return __wbindgen_exn_store;
    },
    get __wbindgen_externrefs () {
        return __wbindgen_externrefs;
    },
    get __wbindgen_free () {
        return __wbindgen_free;
    },
    get __wbindgen_malloc () {
        return __wbindgen_malloc;
    },
    get __wbindgen_realloc () {
        return __wbindgen_realloc;
    },
    get __wbindgen_start () {
        return __wbindgen_start;
    },
    get calculate_state_root () {
        return calculate_state_root;
    },
    get create_proof () {
        return create_proof;
    },
    get decode_proof () {
        return decode_proof;
    },
    get get_runtime_version () {
        return get_runtime_version;
    },
    get memory () {
        return memory;
    },
    get run_task () {
        return run_task;
    },
    get testing () {
        return testing;
    },
    get wasm_bindgen__closure__destroy__hb454d70b27c07de9 () {
        return wasm_bindgen__closure__destroy__hb454d70b27c07de9;
    },
    get wasm_bindgen__convert__closures_____invoke__h0dafe9b2bb334f52 () {
        return wasm_bindgen__convert__closures_____invoke__h0dafe9b2bb334f52;
    },
    get wasm_bindgen__convert__closures_____invoke__h374ab0d2ab71a4a8 () {
        return wasm_bindgen__convert__closures_____invoke__h374ab0d2ab71a4a8;
    }
});
var memory;
var calculate_state_root;
var create_proof;
var decode_proof;
var get_runtime_version;
var run_task;
var testing;
var wasm_bindgen__closure__destroy__hb454d70b27c07de9;
var wasm_bindgen__convert__closures_____invoke__h0dafe9b2bb334f52;
var wasm_bindgen__convert__closures_____invoke__h374ab0d2ab71a4a8;
var __wbindgen_malloc;
var __wbindgen_realloc;
var __wbindgen_exn_store;
var __externref_table_alloc;
var __wbindgen_externrefs;
var __wbindgen_free;
var __wbindgen_start;
